src/tamp_improv/approaches/improvisational/policies/base.py

```diff
@@ -0,0 +1,109 @@
+"""Base policy interface for improvisational approaches."""
+
+import json
+import pickle
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Generic, TypeVar
+
+import gymnasium as gym
+import numpy as np
+from relational_structs import GroundAtom
+
+ObsType = TypeVar("ObsType")
+ActType = TypeVar("ActType")
+
+
+@dataclass
+class TrainingData:
+    """Container for policy training data."""
+
+    states: list[Any]  # List of states where intervention needed
+    preconditions: list[set[GroundAtom]]  # Preconditions to maintain
+    config: dict[str, Any]  # Additional configuration
+
+    def __post_init__(self):
+        """Validate data."""
+        assert len(self.states) == len(
+            self.preconditions
+        ), f"Must be equal length: {len(self.states)}, {len(self.preconditions)})"
+
+    def __len__(self) -> int:
+        return len(self.states)
+
+    def save(self, path: Path) -> None:
+        """Save training data to disk."""
+        path.mkdir(parents=True, exist_ok=True)
+
+        # Save states as numpy array
+        states_path = path / "states.npy"
+        np.save(states_path, np.array(self.states))
+
+        # Save preconditions as pickle (since they contain custom objects)
+        preconditions_path = path / "preconditions.pkl"
+        with open(preconditions_path, "wb") as f:
+            pickle.dump(self.preconditions, f)
+
+        # Save config as JSON
+        config_path = path / "config.json"
+        with open(config_path, "w", encoding="utf-8") as f:
+            json.dump(self.config, f)
+
+    @classmethod
+    def load(cls, path: Path) -> "TrainingData":
```

Comment:
add `from __future__ import annotations` at the top of the file and then you can just do `TrainingData` here